### ISSUE TITLE
runtime: Use static_sandbox_resource_mgmt=true for TEEs

### ISSUE THREAD

**Issue by fidencio** [kata-containers/kata-containers#6818]
When this option is enabled the runtime will attempt to determine the
appropriate sandbox size (memory, CPU) before booting the virtual
machine.

As TEEs do not support memory and CPU hotplug, this approach must be
used.

**Comment by ms-mahuber** [comment]
Cross-referencing the related closed [Issue 117](https://github.com/confidential-containers/documentation/issues/117), I think switching over to static memory management will give us pain in terms of resource allocation (potentially depending on OS/VMM/HV). I would like to list issues I observed and see if it makes sense to open a separate discussion, or if there is already solutions etc. This is not start a lengthy discussion in the PR or to block it as the PR totally makes sense - I am happy to discuss this elsewhere but for now this seems a good spot.

**Concern:** I am concerned about situations in which the view of allocated resources on kubernetes layer (see e.g., `kubectl describe node`) diverges too much from the resources the shim *actually* allocates under the plate - we should avoid/prevent such scenarios as much as possible. Otherwise, terrible pod creation failures can occur.

**Background:** I am thinking in the short term where we use static memory management and where we have not sorted out how to leverage containerd's Sandbox API changes perfectly well. I am mainly writing this regarding CPU and memory resources.

**Definition:** When I speak about **base VM memory/CPU resources** I am referring to `default_vcpus` and `default_memory` see e.g. `src/runtime/config/configuration-clh.toml.in`. These are resources the shim allocates for the base OS in the VM (e.g. kata-agent) but *not* for the workload. Resources dedicated for workloads are 'requested' via kubernetes resource limits, and the kata-shim is aware of those and will add those to the base resources (see e.g. `src/runtime/pkg/oci/utils.go: SandboxConfig()` where a check for static resource management is present, and where resources are accordingly added).

**Potential Issues:**
1. **Potential over-allocation of vCPUs in kata-shim:** In some scenarios, the shim is prone to request too many vCPUs, leading to pod creation faillures.
    - Example with podOverhead of 0.25 and a node with capacity of 3.86 CPU shares (a node with 4 CPUs and a bit of CPU reserved for the host kubernetes runtime itself, specified via e.g. `--kube-reserved=cpu=140m,memory=2662Mi`). If a user requests 3.25 CPU shares via a limit, kubernetes accepts the pod yaml (3.25+0.25<3.86). Under the plate, the shim allocates 1+ceil(3.25) vCPUs = 5 vCPUs as part of the calculations in `src/runtime/pkg/oci/utils.go: SandboxConfig()` (note `default_vcpus = 1` and the ceil of the request is derived from the `updateResources` function. At least with CH this leads to a pod bring-up failure because CH basically has an assert that does not allow more vCPUs than present CPUs.
    - This will need improvement, I think an issue should definitely be opened. While it is plausible to use at least 1 vCPU, adding the limit value this way is bad.
    - Note, that in the dynamic resource management case, this over-allocation can also happen. I am observing warnings where we try to hotplug 4 vCPUs to a 1vCPU VM, but cloud-hypervisor rejects this, and then kata-shim falls back to using 4 vCPUs, i.e., allocates the max number of CPUs available.
2. **Request requests are still not propagated:** While we currently should use resource limits, we should ensure that nobody can specify resource requests. Otherwise, the view of kubernetes on allocated resources can vary drastically from what the shim actually allocates. Imaging you request 100MB of memory and define a limit of 1 GB. The shim will allocate much more memory. While on kubernetes layer, only 100MB are allocated.
    - I think for now this can be documented and solved with a workaround on kubernetes layer: We should document that one can avoid this situation by prohibiting resource requests, for instance via a [validating admission policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/)
3. **When no resource limits are specified, only the base memory/CPU is made available to the VM and no cgroups are enforced in the VM:** This is definitely a problem. Malicious workloads can exhaust the whole UVM resources, and it's also just not nice because we practically reserve no dedicated resources for the workloads, but only for the VM base OS like kata-agent and systemd (where applicable).
    - A potential solution could be to always ensure a default amount of resources for workloads. I tried making a patch on shim level that would introduce new configuration variables which would be used ensure that if no limits are given, a default amount of CPU and memory is allocated for the workloads (see again `src/runtime/pkg/oci/utils.go: SandboxConfig()`). However, this again would lead to a distorted view of the resources Kubernetes thinks are allocatable, and the resources the shim actually allocates under the plate
    - A potential better solution could be to define a default memory/CPU limit on Kubernetes layer view [LimitRanges](https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/). This ensures we can define a good workload memory/CPU value and Kubernetes will be aware of it. Further, we can practically reduce the default base OS memory value of 2,048 MB for the UVM to a much smaller value.
4. **podOverhead in runtimeClasses may not match with default base OS VM memory/CPU allocation made by the shim:** We should align the base memory/CPU assignment for the VM between kata-shim and our podOverhead specification while taking the host memory consumption into account as well.
    - podOverhead values are not translated down to shim level. These are basically an implicit assumption. For each pod, kubernetes substracts allocatable resources by what is specified in podOverhead. This needs to match with what the shim allocates for its base OS (e.g. if you define 1 vCPU for the base VM OS and 256 MB default memory for the base VM OS, this will be part of the podOverhead) **and** with additional resources required on the host, e.g., for starting/running a cloud-hypervisor process.
    - I believe this needs to be addressed in documentation: For instance, I was using cloud-hypervisor with a kata image parameter. In this case CH uses "--pmem file=" and mapped my kata image into private memory mappings. My image was 384 MB large leading to additional 384 MB required per pod just in the CH process. We need to exactly measure how many resources we want to reserve on host side, but also take into account which resources the kata shim will allocated for the base VM OS

I hope this is fruitful input.

**Comment by fidencio** [comment]
@ms-mahuber, first of all, thanks for bringing up the discussion.  I'll try to address all the comments with what I think is the correct approach to take, okay?

There are basically 2 situations we need to take into consideration when we think about a not accurate resource allocation.
1. Allocataing more resources than actually used, which leads to under utilisation of the node;
2. Allocating less resources than actually used, which leads to pod creation failures and whatnot.

I see your comments are more focused on the 2nd case, right?

> **Potential over-allocation of vCPUs in kaa-ta-shim**

This is not a solution, but refining the values we have for the podOverhead is a must for this case.
Right now, for the TEEs case, in the `main` branch, we're defining the podOverhead to match what's configured by default in the configuration.toml, which means an overhead of 1vCPU and 2048MiB of memory, see https://github.com/kata-containers/kata-containers/blob/3a4b9242267ed8578e029cf0bcf10d109f30066e/tools/packaging/kata-deploy/runtimeclasses/kata-runtimeClasses.yaml#L14-L52.

Mind, though, that this is yet not properly set by the Confidential Containers Operator, neither this is part of the `CCv0` branch yet.

This approach, combined with the static resources management, combined with using only limits and not resources, may lead to allocating more memory than used.  However, the next step would be to fine tune the podOverhead values and go with something that would be closer to an "accurate" result.

Issues should be opened for a better estimation of the VMM consumption, leading to updating the values we have for podOverhead.

WDYT about this approach?


> **Request requests are still not propagated**

I cannot agree more to a statement. :-)

> ** When no resource limits are specified, only the base memory/CPU is made available to the VM and no cgroups are enforced in the VM ((

I agree that this is a problem, and I'd like us to explore what can be done with your second suggestion.  However, this is still only something we can advise people to use, document it well, but I don't see a proper way to enforce it.

This deserves an issue on its own.  Would you mind opening it?

> **podOverhead in runtimeClasses may not match with default base OS VM memory/CPU allocation made by the shim**

I think I covered this as part of the reply for the first item.  Please, let me know if you think I didn't, and we can expand more on this.

In short, I agree.


With everything said, I still believe that those problems do exist and should be solved regardless of using the static resource management, and that using it is the way to go.

**Comment by ms-mahuber** [comment]
Thanks for the quick reply Fabiano and for going thoroughly through my comment. Let's proceed as you suggested, agree with you. I will follow up on the points I brought up in separate issues.
